src/SAML2/XML/saml/AttributeValue.php

```diff
@@ -155,21 +152,28 @@ public function toXML(DOMElement $parent = null): DOMElement
         switch ($type) {
             case "integer":
                 // make sure that the xs namespace is available in the AttributeValue
+                $e->setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:xsi', C::NS_XSI);
                 $e->setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:xs', C::NS_XS);
                 $e->setAttributeNS(C::NS_XSI, 'xsi:type', 'xs:integer');
                 $e->textContent = strval($this->getValue());
                 break;
             case "NULL":
+                $e->setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:xsi', C::NS_XSI);
                 $e->setAttributeNS(C::NS_XSI, 'xsi:nil', '1');
                 $e->textContent = '';
                 break;
-            case "array":
-                foreach ($this->getValue() as $object) {
-                    $object->toXML($e);
+            case "object":
+                $value = $this->getValue();
+                if ($value instanceof Chunk) {
+                    // The object is unknown to this library - add an xsi:type for it
+                    $e->setAttributeNS('http://www.w3.org/2000/xmlns/', 'xmlns:xsi', C::NS_XSI);
+                    $e->setAttributeNS(C::NS_XSI, 'xsi:type', $value->getQualifiedName());
```

Comment:
Is the qualified name appropriate as `xsi:type`? I'm basically wondering if we should try to use the `xsi:type` coming from the element itself.

Also, I totally forgot, but does `getQualifiedName()` get you the name with the full namespace, or just the prefix?